### PR TITLE
Use DI for CurrentUserLoader

### PR DESCRIPTION
**Pull request by fraxachun**

### CODE REVIEW

File: packages/api/cms-api/src/user-permissions/user-permissions.module.ts [hunk 1]
```diff
@@ -25,15 +25,27 @@ export class UserPermissionsModule {
...
     static forRootAsync(asyncOptions: UserPermissionsAsyncOptions): DynamicModule {
         return {
             module: UserPermissionsModule,
             imports: asyncOptions.imports,
-            providers: [this.createProvider(asyncOptions)],
+            providers: [
+                this.createProvider(asyncOptions),
+                {
+                    provide: USER_PERMISSIONS_SERVICE,
+                    useClass: UserPermissionsService,
```
Comment (nsams) [inline packages/api/cms-api/src/user-permissions/user-permissions.module.ts:45]: You are exporting UserPermissionsService twice I think

1. here, under the key USER_PERMISSIONS_SERVICE
2. in line 16 under the key UserPermissionsService

I think there should only be 1.

File: packages/api/cms-api/src/auth/strategies/auth-proxy-jwt.strategy.ts [hunk 2]
```diff
@@ -45,12 +47,11 @@ export function createAuthProxyJwtStrategy({
                 audience,
                 ...strategyOptions,
             });
+            if (!this.currentUserLoader) this.currentUserLoader = currentUserLoader ?? new CurrentUserLoader();
```
Comment (nsams) [inline packages/api/cms-api/src/auth/strategies/auth-proxy-jwt.strategy.ts:50]: I think the default should be defined in the module config, would that be possible? Or no default at all to make it more explicit. Or is the default a common use case?

Currently you define it in AppModule, shouldn't that be part of a module?

File: packages/api/cms-api/src/user-permissions/user-permissions.module.ts (hunk 1 above)
Comment (fraxachun) [reply packages/api/cms-api/src/user-permissions/user-permissions.module.ts]: 763cfa12af87d57ad52bf16d26bc5fd3b1f531b2

File: packages/api/cms-api/src/auth/strategies/auth-proxy-jwt.strategy.ts (hunk 2 above)
Comment (fraxachun) [reply packages/api/cms-api/src/auth/strategies/auth-proxy-jwt.strategy.ts]: 42dc30d0f8c2f8ef996358c1efd0c1e8b0f63dd9
The CurrentUserLoader now has a more specific interface and is optional, it should be only set when it is required to really load the currentUser (e.g. when using the UserPermissions-Module).

File: packages/api/cms-api/src/user-permissions/user-permissions.module.ts (hunk 1 above)
Comment (nsams) [reply packages/api/cms-api/src/user-permissions/user-permissions.module.ts]: You are still exporting twice: Line 17 vs. line 23

(you "only" removed the (minimal) code duplication)

Comment (nsams) [reply packages/api/cms-api/src/user-permissions/user-permissions.module.ts]: why export it as `USER_PERMISSIONS_SERVICE` at all, using just `UserPermissionsService` (as before) should be fine? (as it's not injectable - or is that planned?)

File: packages/api/cms-api/src/user-permissions/user-permissions.module.ts [hunk 3]
```diff
@@ -6,15 +6,24 @@ import { UserPermission } from "./entities/user-permission.entity";
...
     imports: [MikroOrmModule.forFeature([UserPermission, UserContentScopes])],
-    providers: [UserPermissionsService, UserResolver, UserPermissionResolver, UserContentScopesResolver, UserPermissionsService],
-    exports: [UserPermissionsService],
+    providers: [
+        UserPermissionsService,
+        UserResolver,
+        UserPermissionResolver,
+        UserContentScopesResolver,
+        {
+            provide: USER_PERMISSIONS_SERVICE,
```
Comment (johnnyomair) [inline packages/api/cms-api/src/user-permissions/user-permissions.module.ts:22]: Why use a constant at all? You could also provide the service using default shorthand syntax for services? https://docs.nestjs.com/fundamentals/custom-providers#standard-providers

File: demo/api/src/auth/auth.module.ts [hunk 4]
```diff
@@ -1,21 +1,26 @@
...
 import { CurrentUser } from "./current-user";
+import { staticUsers } from "./static-users";
 import { UserService } from "./user.service";
 
 @Module({
     providers: [
+        {
+            provide: CURRENT_USER_LOADER,
+            useClass: UserPermissionsCurrentUserLoader,
+        },
```
Comment (johnnyomair) [inline demo/api/src/auth/auth.module.ts:20]: This is the default case, if there's not custom user loader? Can't the module provide it in the library itself?

File: packages/api/cms-api/src/auth/strategies/static-authed-user.strategy.ts [hunk 5]
```diff
@@ -1,27 +1,32 @@
...
-import { ExtractJwt, Strategy } from "passport-jwt";
+import { Strategy } from "passport-custom";
 
 import { CurrentUserInterface } from "../current-user/current-user";
+import { CURRENT_USER_LOADER, CurrentUserLoaderInterface } from "../current-user/current-user-loader";
 
 interface StaticAuthedUserStrategyConfig {
-    staticAuthedUser: CurrentUserInterface;
+    staticAuthedUser?: CurrentUserInterface;
+    staticAuthedUserId?: string;
```
Comment (johnnyomair) [inline packages/api/cms-api/src/auth/strategies/static-authed-user.strategy.ts:10]: We could also achieve this with `staticAuthedUser: CurrentUserInterface | string` and a typeof check. Would remove the edge case where neither option is given.

File: demo/api/src/auth/auth.module.ts (hunk 4 above)
Comment (fraxachun) [reply demo/api/src/auth/auth.module.ts]: Good idea and it solves all other issues as well!

**Review by nsams — APPROVED** [review]

**Review by johnnyomair — APPROVED** [review]

**Pull request merged**